source/core_mqtt_serializer.c

```diff
@@ -357,95 +588,1308 @@ static MQTTStatus_t checkPublishRemainingLength( size_t remainingLength,
+
+                if( *pPropertyLength < pUserProperty->keyLength )
+                {
+                    status = MQTTMalformedPacket;
+                }
+                else
+                {
+                    pVariableHeader = &pVariableHeader[ sizeof( uint16_t ) ];
+                    pUserProperty->pKey = ( const char * ) pVariableHeader;
+                    *pPropertyLength -= pUserProperty->keyLength;
+                    pVariableHeader = &pVariableHeader[ pUserProperty->keyLength ];
+
+                    if( *pPropertyLength < sizeof( uint16_t ) )
+                    {
+                        status = MQTTMalformedPacket;
+                    }
+                    else
+                    {
+                        pUserProperty->valueLength = UINT16_DECODE( pVariableHeader );
+                        *pPropertyLength -= sizeof( uint16_t );
+                        pVariableHeader = &pVariableHeader[ sizeof( uint16_t ) ];
+
+                        if( *pPropertyLength < ( size_t ) ( pUserProperty->valueLength ) )
+                        {
+                            status = MQTTMalformedPacket;
+                        }
+                        else
+                        {
+                            pUserProperty->pValue = ( const char * ) pVariableHeader;
+                            pVariableHeader = &pVariableHeader[ pUserProperty->valueLength ];
+                            *pPropertyLength -= pUserProperty->valueLength;
+                        }
+                    }
+                }
+            }
+
+            *pIndex = pVariableHeader;
+
+            if( ( *count == ( uint32_t ) MAX_USER_PROPERTY ) && ( status == MQTTSuccess ) )
+            {
+                LogDebug( ( "Discarded additional user property with key %s and value %s",
+                            discardUserProperty.pKey,
+                            discardUserProperty.pValue ) );
+            }
+            else
+            {
+                *count += 1U;
+            }
+
+            return status;
+        }
+
+    #endif /* if ( MQTT_USER_PROPERTY_ENABLED ) */
+
+    static MQTTStatus_t MQTT_GetAuthInfoSize( const MQTTAuthInfo_t * pAuthInfo,
+                                              size_t * pPropertyLength )
+    {
+        MQTTStatus_t status = MQTTSuccess;
+
+        /*Validate the parameters*/
+        if( ( pAuthInfo->authMethodLength == 0U ) && ( pAuthInfo->authDataLength != 0U ) )
+        {
+            status = MQTTBadParameter;
+        }
+        else if( ( pAuthInfo->authMethodLength != 0U ) && ( pAuthInfo->pAuthMethod == NULL ) )
+        {
+            status = MQTTBadParameter;
+        }
+        else if( ( pAuthInfo->authDataLength != 0U ) && ( pAuthInfo->pAuthData == NULL ) )
+        {
+            status = MQTTBadParameter;
+        }
+        else
+        {
+            /*Add authentication method and data if provided*/
+            if( pAuthInfo->authMethodLength != 0U )
```

Comment:
We have already checked authMethodLength & authDataLength != 0 above, we can remove these checks here